torch/csrc/autograd/forward_grad.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+
+namespace torch { namespace autograd {
+
+struct ForwardGrad;
+
+// This file contains two classes that are used to store forward AD gradients and
+// ensure that they are scoped properly.
+// Because forward AD runs concurently with the evaluation of the function, we need
+// a mechanism to separate different forward AD invocations and be able to compute the
+// right gradients. We model such invocations as levels here.
+// The particular scoping issue mentionned above has two main drive:
+//   - Ensure that we can conveniently use forward AD within a high level API without
+//     leaking the forward AD states outside.
+//   - Ensure that we can keep the level that we expose to the user API simple (an integer
+//     that represents the nesting depth) while avoiding confusions when the level index
+//     is re-used.
+
+// The important external APIs from this file are:
+//   - ForwardADLevel::get_next_idx() that can be used to enter a new level and get its index
+//   - ForwardADLevel::release_idx() that can be used to exit a given level.
+//   - ForwardGrad() can be used to store a given forward gradient that will handle the level
+//     tracking automatically.
+
+// The behavior above is achieved by ensuring that the ForwardGrad containing gradients for a
+// given level register themselves properly with the corresponding level.
+// On the other hand, the level, when it is released, will reset all the gradients for this
+// level on all the ForwardGrad.
```

Comment:
?

Reply:
What's the basic implementation strategy?

Reply:
"Every tensor has a ForwardGrad, maintaining a map from levels to tangents. ForwardGrad has a mutex, so that we can safely remove tensors from the map when level is exited."

Reply:
I also expect some discussion on the AutogradMeta finalizers

Reply:
I also expect some discussion about why the locking mechanism here is correct